075-auto-rebalancing-cluster-scaling.md

```diff
@@ -0,0 +1,368 @@
+
+1. On the **1st** reconciliation (scaling down is skipped, rebalancing started):
+     * the check on scaling down the cluster fails, but it doesn't have to revert the `replicas` back because the `Kafka.spec.cruiseControl.autoRebalance` is provided. So the reconciliation doesn't fail.
+     * the `scaleDown()` method skips doing the scale down because a rebalancing is needed first (and the check failed).
+     * the `autoreRalancing()` method detects that a scaling down was requested because of `KafkaCluster.removedNodes()` returning a non empty list with the brokers' IDs to be removed from the cluster. Because the `Kafka.status.cruiseControl.autoRebalance.state` doesn't exist, it's a new operation so a new `KafkaRebalance` custom resource is created and the `Kafka.status.cruiseControl.autoRebalance.state` is updated with the corresponding state (should be `ProposalPending`) and the `remove-brokers` with the brokers' IDs.
+2. From the **2nd** to the **(N-1)th** reconciliation (rebalancing is going on, checks when it's done):
+     * the checks on scaling down the cluster keeps failing but it doesn't have to revert the `replicas` back because the `Kafka.spec.cruiseControl.autoRebalance` is provided. So the reconciliation doesn't fail.
+     * the `scaleDown()` method skips doing scale down because a rebalancing is needed first (and the check failed).
+     * the `autoRebalancing()` has to check the status of the rebalancing by looking at the corresponding `KafkaRebalance` custom resource status which will move through `ProposalPending`, `ProposalReady` to `Rebalancing` (we are assuming auto-approval). The rebalancing status is reflected into the `Kafka.status.cruiseControl.autoRebalance.state` field.
+3. On the **Nth** reconciliation (rebalancing id done, scaling down can start):
+     * the checks on scaling down the cluster doesn't fail, because the auto-rebalancing was completed and the nodes to be removed don't host any partition replicas anymore.
+     * the `scaleDown()` method can run the scaling down.
+     * the `autoRebalancing()` detects that the `KafkaRebalance` custom resource status reports the `Ready` state (so the rebalancing is completed).
+
+During the above flow, errors can happen before the scaling down:
+
+* There is an error when getting the proposal.
+* After auto-approving the proposal, there is an error when running the rebalance.
+
+In both scenarios, the reconciliation ends with the `KafkaRebalance` custom resource in the `NotReady` state which is reflected into the `Kafka.status.cruiseControl.autoRebalance.state` field.
+
+The reconcile ends but the scale down cannot proceed because of the brokers to be removed are still hosting some partition replicas.
+
+On the next reconciliation, the cluster operator will try to do a scale down again, so running the flow as before.
+The previous `KafkaRebalance` custom resource in the `NotReady` state is overridden to generate a new optimization proposal.
+
```

Comment:
It would be great to include some more scenarios especially around various error situations. E.g.,: 
* what happens if I scale-down and up in a rapid sequence before the rebalancing is completed? 
* How does this deal with CC restarts? I raised it before -> but the scale-up will actually restart CC right after adding the nodes and likely kill the rebalance in the current flow.

Reply:
I updated the PR in that regards.